middlewares/studentAssessmentAccess.js

```diff
@@ -0,0 +1,168 @@
+//        checkBrowserExamKeys(res, req.headers['x-safeexambrowser-requesthash']);
+//    }
+
+    // SEB: check user-agent
+    if ('user-agent' in req.headers) {
+        checkUserAgent(res, req.headers['user-agent']);
+    }
+
+// Future feature
+//    // Course Instance view (only show allowed exam)
+//    if (!('assessment' in res.locals)) {
+//        //console.log('In Course Instance Page');
+//        //res.locals.authz_data.mode = 'SEB';
+//        //console.log(res.locals);
+//    }
+
+    // SafeExamBrowser protect the assesment
+    if ('authz_result' in res.locals &&
+        res.locals.authz_result.mode == 'SEB') {
+
+        // If the assessment is complete, use this middleware to show the logout page
+        if ('assessment_instance' in res.locals && res.locals.assessment_instance.open == false) {
+            return badSEB(req, res);
+        }
+
+        // If any of our auth checks didn't pass, fail (send to download)
+        if (res.locals.authz_data.mode != 'SEB') {
+            return badSEB(req, res);
+        }
+    }
+
+
+    // Password protect the assessment
+    if ('authz_result' in res.locals &&
+        'password' in res.locals.authz_result &&
+        res.locals.authz_result.password) {
+
+        // No password yet case
+        if (req.cookies.pl_assessmentpw == null) {
+            return badPassword(res, '');
+        }
+
+        // Invalid or expired password case
+        var pwData = csrf.getCheckedData(req.cookies.pl_assessmentpw, config.secretKey,
+                                         {maxAge: timeout * 60 * 60 * 1000});
+        if (pwData === null
+            || pwData.password !== res.locals.authz_result.password) {
+            return badPassword(res, 'Password invalid or expired, please try again.');
+        }
+
+        // Successful password case: falls through
+    }
+
+    // Pass-through for everything else
+    next();
+});
+
+router.post('/', function(req, res, next) {
+
+    // For password protected things:
+    if ('authz_result' in res.locals
+        && 'password' in res.locals.authz_result
+        && res.locals.authz_result.password) {
+
+        if (req.body.__action == 'assessmentPassword') {
+
+            if (req.body.password == res.locals.authz_result.password) {
+                var pwCookie = csrf.generateToken({password: req.body.password}, config.secretKey);
+                res.cookie('pl_assessmentpw', pwCookie);
+                return res.redirect(req.originalUrl);
+            } else {
+                return badPassword(res);
+            }
+        }
+    }
+
+    // Fallthrough for the middleware
+    next();
+});
+module.exports = router;
+
+function badPassword(res, msg) {
+
+    logger.info(`invalid password attempt for ${res.locals.user.uid}`);
+    res.clearCookie('pl_assessmentpw');
+    res.locals.passwordMessage = msg;
+    res.locals.prompt = 'password';
+    return res.render(__filename.replace(/\.js$/, '.ejs'), res.locals);
```

Comment:
Rather than hijacking the route and rendering a page in a middleware, can we use the same pattern as we do for authn where we redirect to a page requesting the password and put the originalUrl in a query parameter?

Reply:
Yeah, that sounds very reasonable. It felt weird making an .ejs for middleware.